shared_stack.bash

```diff
@@ -0,0 +1,214 @@
+# latest versions of the LSST code should prefer the lsstsw build tool.
+#
+# Specifically, we:
+#
+# - Get the latest version of the lsstinstall tool into the STATE directory;
+# - Get the list of tags from the EUPS distribution server;
+# - For each tag which is not in the list of already-installed-and-deleted tags
+#   in "tags.deleted" and which does not have a directory under $ROOT/tag:
+#   - use lsstinstall to install the appropriate rubin-env conda environment if
+#     needed
+#   - add useful additional packages for developers if necessary
+#   - use "eups distrib install" to install the Science Pipelines from binary
+#     artifacts for the top-level PRODUCT
+#   - create a loadLSST.sh script for users to load the environment from
+#     scratch, a switch.sh script to activate the environment if conda is
+#     already loaded, and an env_name file to simplify other tooling that may
+#     need to map tags to rubin-env versions;
+# - Maintain symlinks for the latest daily and weekly tags;
+# - Maintain a common base loadLSST.sh script;
+# - For each tag beyond the number to be kept for dailies (KEEP_d) and
+#   weeklies (KEEP_w):
+#   - add the tag to the "tags.deleted" list
+#   - temporarily move the $ROOT/tag subdirectory aside
+#   - untag (if shared) or remove (if exclusive) the eups packages with the tag
+#   - remove the $ROOT/tag subdirectory
+#
+# A -n option may be used to do a "dry run" that only prints out the commands
+# that would be executed.
+#
+# This code is intended to work properly if invoked multiple times in parallel,
+# e.g. if it takes longer than the interval between cron jobs, although there
+# is no resulting speedup.  Hence PID-labeled temporary directories and atomic
+# moves are used frequently.
+
+###############
+# CONFIGURATION
+###############
+
+# Software root directory
+ROOT=/scratch/rubinsw
+# State directory
+STATE=~/shared-stack
+# Top-level product to install
+PRODUCT=lsst_distrib
+# Number of daily releases to keep
+# shellcheck disable=SC2034
+KEEP_d=24
+# Number of weekly releases to keep
+# shellcheck disable=SC2034
+KEEP_w=26
+
+#######
+# Setup
+#######
+
+while getopts n opt; do
+  case "$opt" in
+    n) dryrun="echo"
+       ;;
+    *) echo "Usage: $0 [-n]"
+       exit 1 ;;
+  esac
+done
+[ $OPTIND -le $# ] && echo "Usage: $0 [-n]" && exit 1
+# Ensure that there's no contamination from a pre-existing environment
+unset LSST_CONDA_ENV_NAME PYTHONPATH
+
+##############
+# Installation
+##############
+
+set -e
+mkdir -p "$STATE"
+cd "$STATE"
+# Label temporaries with our unique PID
+tmp=tmp.$$
+mkdir -p "$tmp" "$ROOT/tag"
+[ -e tags.deleted ] || touch tags.deleted
+umask 022
+
+# Get list of tags
+$dryrun curl -L https://ls.st/lsstinstall -o $tmp/lsstinstall
+$dryrun mv $tmp/lsstinstall lsstinstall
+curl -L https://eups.lsst.codes/stack/src/tags | while read -r line; do
+  [[ "$line" =~ \>([dvw][0-9_]+(rc[0-9]+)?)\.list\< ]] && echo "${BASH_REMATCH[1]}"
+done | sort > $tmp/tags
```

Comment:
What I have done is prep a `tags.deleted` list that excludes most old tags.  But I don't see how "things that are know not to work" could happen; each tag is installed in an environment (equivalent to one) in which it _was_ known to work.

Reply:
Where is the mapping of lsstinstall (and thus environments) kept? It looks like you are just always curling the latests?

Reply:
I see that as prehaps in a new location trying to install say w_2022_01 with the latest lsstinstall, which would give the latest environment right?

Reply:
No!  Unlike newinstall, lsstinstall is designed to work with any environment; you can always safely use the latest version of lsstinstall for any stack version.  The environment is selected by the tag using the `-T` option.

Reply:
The mapping of tags to environments is kept permanently in eups.lsst.codes.

Reply:
Way outside the scope of this ticket's discussion, but a LONG time ago I provided an eups package that would turn on the correct conda environment when you setup a tag, why did we never go down that direction?

Reply:
Feel free to answer, or ignore so you can move on, and we can catch up some other time about it

Reply:
IIRC, there were objections to having the "higher-level" packaging system swapping out the "underlying" conda environment infrastructure.  We ended up resolving this by conda-packaging eups so that eups stacks could be embedded in conda environments.

There are a few tricky things and misfeatures as a result, but it seems to be working well enough for now.  This could be reconsidered in the future.